rfcs/20190116-embedding-partitioned-variable.md

```diff
@@ -0,0 +1,498 @@
+| **Author** | Yuefeng Zhou (yuefengz@google.com)                                        |
+|**Sponsors**| Alexandre Passos (apassos@google.com), Martin Wicke (wicke@google.com), Josh Levenberg  (joshl@google.com) |
+|**Updated** | 2019-1-16                                                                 |
+
+
+Summary: this RFC describes the design and interfaces for embeddings and partitioned variables in TF 2.0 based on Distribution Strategy. We propose to support embeddings and sharded layers with DistributionStrategy and Keras' layer API. We also propose to shard embeddings and layers only in `ParameterServerStrategy` and mirror them in `MirroredStrategy` or `CollectiveAllReduceStrategy.` 
+
+
+## Background
+
+
+### Embeddings
+
+
+An [embedding](https://www.tensorflow.org/guide/embedding) is a mapping from discrete objects, such as words, to vectors of real numbers. In TensorFlow implementation, we use a variable to hold all these vectors for a feature column. All vectors in an embedding variable have the same size but the sizes of embedding variables may vary. 
+
+We don't always use the whole embedding variable in a training step, the result of embedding lookup is <code>[IndexedSlices](https://www.tensorflow.org/api_docs/python/tf/IndexedSlices)</code> which is essentially a dict mapping from indices to vectors. So is the gradient to an embedding variable.
+
+
+#### Sharded Embedding
+
+
+Sharded embedding is one use case of `PartitionedVariable`. People sometimes shard embedding variables across multiple parameter servers. There are many reasons why people shard embeddings:
+
+1.  parallelize computations : we can do embedding lookups in parallel and at the same time keep communication between machines small; 
+1.  balance parameter servers' load: if an embedding layer is large and placed on one ps, that ps can become a hotspot;
+1.  shard storage: occasionally embedding variables can be too large to fit in the memory of one machine.
+1.  reduce contentions in asynchronous training: updates from different workers could possibly end up on different partitions which reduces possibility of overriding each other's update and using locking, although this depends on the implementation of embedding update.
+
+
+##### Partition Strategy: div or mod
+
+
+There are two strategies to look up embedding vectors on a sharded embedding variable: "div" and "mod". See an explanation [here](https://www.tensorflow.org/api_docs/python/tf/nn/embedding_lookup). "Mod" strategy may be useful when references to embedding slices are not evenly distributed over their indices.  However, "mod" strategy is a poor approximation to the actual load balancing users want. For example, when vocabulary is sorted by frequency, with "mod" strategy the first parameter server always has larger load than the second one. Furthermore, its current checkpointing mechanism also prevents users from migrating to a cluster with different number of parameter servers.
```

Comment:
Would it be possible to support user-provided hash function, or a step further, consistent hashing? It enables PS level 1 fault tolerance, and it would be particularly helpful in the case of large-scale embeddings (thus more ps and workers).

Reply:
That sounds more of general feature request for `ParameterServerStrategy` rather than for embeddings. 

Consistent hashing, if I understand it correctly, is more for scenarios that need high availability. When a server goes down, you'll need a mechanism to migrate its key-values to other servers. However, PS only checkpoints once in a while, so you'll need heads-up from container orchestration system in order to prepare variable migration before a PS goes down. You'll also need to tell workers about the configuration change and workers may have to rebuild the whole graph for the re-assignment of variables. This can be something built on top of TensorFlow although I guess a simpler solution is just to restart the whole cluster when a server needs to be removed since high availability is not the most important goal.